File: hummingbird/androidPatientApp/app/src/main/java/net/acumenus/hummingbird/patient/data/PatientStateVocabulary.kt
package net.acumenus.hummingbird.patient.data

/**
 * The patient API transmits stable state codes. Keep their patient-language
 * rendering explicit and contextual instead of exposing a title-cased internal
 * identifier such as `at_risk`. A future translated registry must preserve the
 * codes and replace only this copy.
 */
internal enum class PatientStateDomain {
    SCHEDULE,
    PATHWAY,
    MILESTONE,
    PATHWAY_EVENT,
    GOAL,
    TIMING_CONFIDENCE,
    DISCHARGE_CRITERION,
    ROUNDS_TOPIC,
    PATHWAY_EVENT_CATEGORY,
}

internal object PatientStateVocabulary {
    const val VERSION = "patient-state-vocabulary.v1-draft"

    private val labelsByDomain: Map<PatientStateDomain, Map<String, String>> = mapOf(
        PatientStateDomain.SCHEDULE to mapOf(
            "requested" to "Requested",
            "planned" to "Planned",
            "confirmed" to "Confirmed",
            "in_progress" to "Happening now",
            "completed" to "Completed",
            "delayed" to "Delayed",
            "canceled" to "No longer planned",
        ),
        PatientStateDomain.PATHWAY to mapOf(
            "planned" to "Planned",
            "current" to "Happening now",
            "completed" to "Completed",
            "delayed" to "Delayed",
            "canceled" to "No longer planned",
        ),
        PatientStateDomain.MILESTONE to mapOf(
            "planned" to "Planned",
            "current" to "Happening now",
            "completed" to "Completed",
            "delayed" to "Delayed",
            "canceled" to "No longer planned",
        ),
        PatientStateDomain.PATHWAY_EVENT to mapOf(
            "planned" to "Planned",
            "current" to "Happening now",
            "completed" to "Completed",
            "delayed" to "Delayed",
            "canceled" to "No longer planned",
        ),
        PatientStateDomain.GOAL to mapOf(
            "proposed" to "Being considered",
            "planned" to "Planned",
            "in_progress" to "In progress",
            "completed" to "Completed",
            "paused" to "Paused",
            "canceled" to "No longer planned",
        ),
        PatientStateDomain.TIMING_CONFIDENCE to mapOf(
            "confirmed" to "Confirmed",
            "estimated" to "Estimated",
            "unknown" to "Not yet known",
        ),
        PatientStateDomain.DISCHARGE_CRITERION to mapOf(
            "met" to "Met",
            "pending" to "Still needed",
            "at_risk" to "Needs attention",
        ),
        PatientStateDomain.ROUNDS_TOPIC to mapOf(
            "discussed" to "Discussed",
            "current" to "Being reviewed",
            "planned" to "Planned",
        ),
        PatientStateDomain.PATHWAY_EVENT_CATEGORY to mapOf(
            "test" to "Test",
            "procedure" to "Procedure",
            "transport" to "Transportation",
            "other" to "Care update",
        ),
    )

    fun label(code: String, domain: PatientStateDomain): String =
        labelsByDomain[domain]?.get(code) ?: "Status being confirmed"

    fun labels(domain: PatientStateDomain): Map<String, String> =
        labelsByDomain.getValue(domain)

    /**
     * Older servers omit this additive field. Preserve that compatibility, but
     * do not render a projection explicitly stamped with unknown state copy.
     */
    fun isCompatible(serverVersion: String?): Boolean =
        serverVersion == null || serverVersion == VERSION
}

File: hummingbird/iosPatientApp/HummingbirdPatient/Models/PatientStateVocabulary.swift
import Foundation

/// The patient API transmits stable state *codes*. Keep their English display
/// copy explicit and contextual here instead of turning internal identifiers
/// such as `at_risk` into title-cased text at render time. A future localized
/// registry must preserve the codes and replace only this approved copy.
enum PatientStateDomain: CaseIterable {
    case schedule
    case scheduleCategory
    case pathway
    case milestone
    case pathwayEvent
    case goal
    case timingConfidence
    case dischargeCriterion
    case roundsTopic
}

enum PatientStateVocabulary {
    static let version = "patient-state-vocabulary.v2-draft"

    private static let labels: [PatientStateDomain: [String: String]] = [
        .schedule: [
            "requested": "Requested",
            "planned": "Planned",
            "scheduled": "Scheduled",
            "confirmed": "Confirmed",
            "waiting": "Waiting",
            "transport_requested": "Transportation requested",
            "in_progress": "Happening now",
            "completed": "Completed",
            "result_pending": "Result not available yet",
            "result_released": "Result available",
            "delayed": "Delayed",
            "canceled": "No longer planned",
        ],
        .scheduleCategory: [
            "test": "Test",
            "procedure": "Procedure",
            "transport": "Transportation",
            "other": "Care update",
        ],
        .pathway: [
            "planned": "Planned",
            "current": "Happening now",
            "completed": "Completed",
            "delayed": "Delayed",
            "canceled": "No longer planned",
        ],
        .milestone: [
            "planned": "Planned",
            "current": "Happening now",
            "completed": "Completed",
            "delayed": "Delayed",
            "canceled": "No longer planned",
        ],
        .pathwayEvent: [
            "planned": "Planned",
            "current": "Happening now",
            "completed": "Completed",
            "delayed": "Delayed",
            "canceled": "No longer planned",
        ],
        .goal: [
            "proposed": "Being considered",
            "planned": "Planned",
            "in_progress": "In progress",
            "completed": "Completed",
            "paused": "Paused",
            "canceled": "No longer planned",
        ],
        .timingConfidence: [
            "confirmed": "Confirmed",
            "estimated": "Estimated",
            "unknown": "Not yet known",
        ],
        .dischargeCriterion: [
            "met": "Met",
            "pending": "Still needed",
            "at_risk": "Needs attention",
        ],
        .roundsTopic: [
            "discussed": "Discussed",
            "current": "Being reviewed",
            "planned": "Planned",
        ],
    ]

    static func label(for code: String, domain: PatientStateDomain) -> String {
        labels[domain]?[code] ?? "Status being confirmed"
    }

    static func labels(for domain: PatientStateDomain) -> [String: String] {
        labels[domain] ?? [:]
    }

    /// Older servers did not advertise this additive value. Preserve their
    /// existing compatibility, but never render a projection explicitly
    /// stamped with an unfamiliar state-language vocabulary.
    static func isCompatible(serverVersion: String?) -> Bool {
        serverVersion == nil || serverVersion == version
    }
}
